pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -791,9 +799,27 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 		nodes = append(nodes, node)
 	}
 
-	// Sort the list of nodes by how many devices they contain.
+	getNodePriority := func(n int) nodePriority {
+		if hint.NUMANodeAffinity.IsSet(n) {
+			return priorityFromAffinity
+		}
+
+		if n != nodeWithoutTopology {
+			return priorityNotFromAffinity
+		}
+
+		return priorityWithoutTopology
+	}
+
+	// Sort the list of nodes by nodes from affinity set, nodes from none-affinity
+	// set, nodes without topology set and how many devices they contain.
 	sort.Slice(nodes, func(i, j int) bool {
-		return perNodeDevices[i].Len() < perNodeDevices[j].Len()
+		pi, pj := getNodePriority(nodes[i]), getNodePriority(nodes[j])
+		if pi != pj {
+			return pi < pj
+		}
+
+		return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()
```

Comment:
I was thinking something more like:
```
// Sort the list of nodes by:
// 1) Nodes contained in the 'hint's affinity set
// 2) Nodes not contained in the 'hint's affinity set
// 3) The the fake NUMANode of -1 (assuming it is included in the list)
// Within each of the groups above, sort the nodes by how many devices they contain
sort.Slice(nodes, func(i, j int) bool {
	// If one or the other of nodes[i] or nodes[j] is in the 'hint's affinity set
	if hint.NUMANodeAffinity.IsSet(nodes[i]) && hint.NUMANodeAffinity.IsSet(nodes[j]) {
		return nodes[i] < nodes[j]
	}
	if hint.NUMANodeAffinity.IsSet(nodes[i]) {
		return true
	}
	if hint.NUMANodeAffinity.IsSet(nodes[j]) {
		return false
	}
	
	// If either nodes[i] or nodes[j] are in the 'hint's affinity set (but are not -1)
	if nodes[i] != nodeWithoutTopology && nodes[j] != nodeWithoutTopology {
		return nodes[i] < nodes[j]
	}
	if nodes[i] != nodeWithoutTopology {
		return true
	}
	if nodes[j] != nodeWithoutTopology {
		return false
	}
	
	// If both nodes[i] and nodes[j] are -1
	return nodes[i] < nodes[j]
}
```

Reply:
// Within each of the groups above, sort the nodes by how many devices they contain
should this line of code `return nodes[i] < nodes[j]` be something like `return perNodeDevices[nodes[i]].Len() < perNodeDevices[nodes[j]].Len()`